contracts/transfer/src/state.rs

```diff
@@ -223,17 +231,131 @@ impl TransferState {
         self.var_crossover = tx.crossover;
         self.var_crossover_addr.replace(*tx.fee.stealth_address());
 
-        let mut result = Ok(Vec::new());
+        self.execute(tx)
+    }
+
+    /// Executes the contract call if present.
+    ///
+    /// # Panics
+    /// Any failure in the checks performed in processing will result in a
+    /// panic. The contract expects the environment to roll back any change
+    /// in state.
+    pub fn execute(
+        &mut self,
+        tx: Transaction,
+    ) -> Result<Vec<u8>, ContractError> {
+        let mut result = Ok(rusk_abi::RawResult::empty());
 
         if let Some((contract_id, fn_name, fn_args)) = tx.call {
-            result = rusk_abi::call_raw(
-                ContractId::from_bytes(contract_id),
-                &fn_name,
-                &fn_args,
+            let contract_id = ContractId::from_bytes(contract_id);
+            self.gas_price = Some(tx.fee.gas_price);
+            result = rusk_abi::call_raw(contract_id, &fn_name, &fn_args);
+            self.gas_price = None;
+            if let Ok(RawResult {
+                data: _,
+                economic_mode,
+            }) = result.clone()
+            {
+                match economic_mode {
+                    EconomicMode::Charge(charge) if charge != 0 => self
+                        .apply_charge(&contract_id, charge, tx.fee.gas_price),
+                    EconomicMode::Allowance(allowance) if allowance != 0 => {
+                        self.apply_allowance(
+                            &contract_id,
+                            allowance,
+                            tx.fee.gas_price,
+                        )
+                    }
+                    _ => (),
+                }
+            }
+        }
+
+        result.map(|r| r.data)
+    }
+
+    /// Applies contract's charge.
+    /// Caller of the contract will pay a larger fee
+    /// so that contract can earn the difference between charge
+    /// and the actual cost of the call.
+    fn apply_charge(
+        &mut self,
+        contract_id: &ContractId,
+        charge: u64,
+        gas_price: u64,
+    ) {
+        let cost = (rusk_abi::spent() + SURCHARGE_POINTS) * gas_price;
+        if charge > cost {
+            let earning = charge - cost;
+            self.add_balance(*contract_id, earning);
+            rusk_abi::set_charge(earning);
```

Comment:
Following the above discussion on SURCHARGE, I think this should be set to `charge` and let Rusk manage this along with the gas spent.